cpp/src/arrow/compute/exec.cc

```diff
@@ -1034,9 +1034,23 @@ class VectorExecutor : public KernelExecutorImpl<VectorKernel> {
     output_num_buffers_ = static_cast<int>(output_type_.type->layout().buffers.size());
 
     // Decide if we need to preallocate memory for this kernel
-    validity_preallocated_ =
-        (kernel_->null_handling != NullHandling::COMPUTED_NO_PREALLOCATE &&
-         kernel_->null_handling != NullHandling::OUTPUT_NOT_NULL);
+    validity_preallocated_ = false;
+    if (output_type_.type->id() != Type::NA) {
+      if (kernel_->null_handling == NullHandling::COMPUTED_PREALLOCATE) {
+        // Override the flag if kernel asks for pre-allocation
+        validity_preallocated_ = true;
+      } else if (kernel_->null_handling == NullHandling::INTERSECTION) {
+        bool elide_validity_bitmap = true;
+        for (const auto& arg : batch.values) {
+          auto null_gen = NullGeneralization::Get(arg) == NullGeneralization::ALL_VALID;
+
+          // If not all valid, this becomes false
+          elide_validity_bitmap = elide_validity_bitmap && null_gen;
+        }
+        validity_preallocated_ = !elide_validity_bitmap;
+      }
+    }
```

Comment:
I don't think this logic is safe. If `null_handling` is `INTERSECTION` the function might want to follow a single code path that uses the pre-allocated buffer to put the result of the bitmaps intersection.

And even if that is not the case (assuming all kernels are perfect at the moment), this extra logic introduces branching in the possible states of pre-allocated buffers based on runtime input data and not something statically defined in the `kernel_` when it's configured.

If you want the speed benefits of not pre-allocating a validity bitmap when not necessary, declare the kernel as `COMPUTED_NO_PREALLOCATE` and work on the kernel logic to allocate only when needed. The generic binding and execution logic should be simple and not depend on values so much. It's too complicated as it is.

Reply:
> follow a single code path that uses the pre-allocated buffer to put the result of the bitmaps intersection

Are you talking about `PropagateNulls`? Currently, seems this logic is used in `INTERSECTION`.

If not, i understand that if the kernel internal execute path must `follow a single code path that uses the pre-allocated buffer`, `COMPUTED_PREALLOCATE` can be set statically, which should be safer for this kernel?

Back to preallocation-buffer, if kernel-function sets `NullHandling::INTERSECTION`, does it mean that kernel-function may not be sure whether it needs a pre-allocated buffer and let the Executor decide for itself? At this time, the Executor can only dynamically determine whether to set the pre-allocate buffer based on the input data.

> And even if that is not the case (assuming all kernels are perfect at the moment), this extra logic introduces branching in the possible states of pre-allocated buffers based on runtime input data and not something statically defined in the `kernel_` when it's configured.

From a performance perspective, the current logic does introduce many branches at runtime for `INTERSECTION`, compared to `PropagateNulls`, it actually passes `NullGeneralization::Get` introduces very few branches.

Reply:
Besides that, current ScalarExecutor will pre-allocate with `NullGeneralization` in INTERSECTION mode.

Reply:
Sorry, by branches I meant branches in the state-space the execution framework and the kernels might traverse. Meaning that without duplication of test cases for every kernel, we might hide a latent bug until an unlikely array configuration causes a SIGSEGV on the non-pre-allocated validity buffer.

> Back to preallocation-validity buffer, if kernel-function sets `NullHandling::INTERSECTION`, does it mean that kernel-function may not be sure whether it needs a pre-allocated validity buffer and let the Executor decide for itself? At this time, the Executor can only dynamically determine whether to set the pre-allocate buffer based on the input data.

By statically I mean a one-time configuration of the `kernel_` when it's added to the function in the registry.

For a kernel implementer, it's simpler to assume `NullHandling::INTERSECTION` implies pre-allocated validity bitmap buffer. No matter what arrays are passed as input.

Reply:
> For a kernel implementer, it's simpler to assume `NullHandling::INTERSECTION` implies pre-allocated validity bitmap buffer. No matter what arrays are passed as input.

I see, thank you for your explanation @felipecrv .

Do you think the changes in the third commit of this PR are reasonable? If so, I will revert the first two commits. The third commit is mainly related to the pre-allocation of output data-buffer and refactored part of the code, which can bring the benefits mentioned by @mapleFU .

In addition, I will create an issue to track the pre-allocation of validity-buffer in `INTERSECTION` mode, because currently `ScalarExecutor` will do pre-allocation according to `NullGeneralization` in this mode.

Reply:
I think the handling of list-view types in the pre-allocation code is good. The 3rd commit has a mix of things in it.

Reply:
Yes, except for [Large]ListView and chunked-array `NullGeneration` support, the others are some minor refactorings (some DCHECKs are unnecessary for now).